File: src/lib/offline-assets.ts
import { existsSync, mkdirSync, statSync, writeFileSync, readFileSync } from 'fs';
import { join } from 'path';
import { execSync } from 'child_process';
import { CacheStatus } from './types';

const DOCKER_STATIC_URL = 'https://download.docker.com/linux/static/stable/aarch64/docker-27.5.1.tgz';
const CACHE_DIR = join(process.env.HOME || '/tmp', '.t3shield-installer');

function log(msg: string, ...args: unknown[]) {
  console.log(`[offline-assets] ${msg}`, ...args);
}

function ensureCacheDir() {
  if (!existsSync(CACHE_DIR)) {
    mkdirSync(CACHE_DIR, { recursive: true });
    log('Created cache directory: %s', CACHE_DIR);
  }
}

export function getCachePaths() {
  return {
    cacheDir: CACHE_DIR,
    dockerTgz: join(CACHE_DIR, 'docker-static.tgz'),
    dockerDir: join(CACHE_DIR, 'docker-static'),
    firmwareTar: join(CACHE_DIR, 'firmware.tar'),
    firmwareVersion: join(CACHE_DIR, 'firmware-version.txt'),
    firmwareDigest: join(CACHE_DIR, 'firmware-digest.txt'),
  };
}

export function getCacheStatus(): CacheStatus {
  const paths = getCachePaths();
  const dockerReady = existsSync(join(paths.dockerDir, 'docker', 'dockerd'));
  const firmwareReady = existsSync(paths.firmwareTar);
  let firmwareTag: string | null = null;
  if (existsSync(paths.firmwareVersion)) {
    firmwareTag = readFileSync(paths.firmwareVersion, 'utf-8').trim();
  }
  return { dockerBinaries: dockerReady, firmwareImage: firmwareReady, firmwareTag };
}

export async function prepareDockerBinaries(
  onProgress?: (msg: string) => void
): Promise<string> {
  ensureCacheDir();
  const paths = getCachePaths();

  if (existsSync(join(paths.dockerDir, 'docker', 'dockerd'))) {
    log('Docker binaries already cached');
    onProgress?.('Docker binaries cached');
    return paths.dockerDir;
  }

  if (!existsSync(paths.dockerTgz)) {
    log('Downloading Docker static binaries from %s', DOCKER_STATIC_URL);
    onProgress?.('Downloading Docker binaries (60MB)...');
    try {
      execSync(`curl -fSL -o "${paths.dockerTgz}" "${DOCKER_STATIC_URL}"`, {
        stdio: 'pipe',
        timeout: 5 * 60 * 1000,
      });
      const size = statSync(paths.dockerTgz).size;
      log('Downloaded Docker binaries: %dMB', Math.round(size / 1024 / 1024));
    } catch (err) {
      throw new Error(`Failed to download Docker binaries: ${err instanceof Error ? err.message : err}`);
    }
  }

  log('Extracting Docker binaries...');
  onProgress?.('Extracting Docker binaries...');
  mkdirSync(paths.dockerDir, { recursive: true });
  try {
    execSync(`tar xzf "${paths.dockerTgz}" -C "${paths.dockerDir}"`, { stdio: 'pipe' });
  } catch (err) {
    throw new Error(`Failed to extract Docker binaries: ${err instanceof Error ? err.message : err}`);
  }

  if (!existsSync(join(paths.dockerDir, 'docker', 'dockerd'))) {
    throw new Error('Docker binaries extraction failed — dockerd not found');
  }

  log('Docker binaries ready at %s', paths.dockerDir);
  onProgress?.('Docker binaries ready');
  return paths.dockerDir;
}

/**
 * Get the remote digest for an image without pulling it.
 * Uses `docker manifest inspect` which only fetches metadata.
 */
function getRemoteDigest(image: string): string | null {
  try {
    const output = execSync(
      `docker manifest inspect "${image}" 2>/dev/null | grep -o '"digest":"[^"]*"' | head -1 | cut -d'"' -f4`,
      { stdio: 'pipe', timeout: 30000 }
    ).toString().trim();
    return output || null;
  } catch {
    return null;
  }
}

export async function prepareFirmwareImage(
  image: string,
  ghcrUsername: string,
  ghcrToken: string,
  onProgress?: (msg: string) => void
): Promise<string> {
  ensureCacheDir();
  const paths = getCachePaths();

  // Check docker is available
  try {
    execSync('docker --version', { stdio: 'pipe' });
  } catch {
    throw new Error('Docker is not installed on this machine. Install Docker and try again.');
  }

  // Login to GHCR (needed for digest check + pull)
  log('Logging in to ghcr.io...');
  onProgress?.('Logging in to container registry...');
  try {
    execSync(`echo "${ghcrToken}" | docker login ghcr.io -u "${ghcrUsername}" --password-stdin`, {
      stdio: 'pipe',
      timeout: 30000,
    });
  } catch {
    throw new Error('GHCR login failed — check username and token in Settings');
  }

  // Check if cache is up to date by comparing digests
  if (existsSync(paths.firmwareTar) && existsSync(paths.firmwareDigest)) {
    const cachedDigest = readFileSync(paths.firmwareDigest, 'utf-8').trim();
    onProgress?.('Checking for firmware updates...');
    const remoteDigest = getRemoteDigest(image);

    if (remoteDigest && cachedDigest === remoteDigest) {
      const size = statSync(paths.firmwareTar).size;
      log('Firmware cache is up to date: %s (%dMB)', remoteDigest, Math.round(size / 1024 / 1024));
      onProgress?.(`Firmware image up to date (${Math.round(size / 1024 / 1024)}MB)`);
      return paths.firmwareTar;
    }

    if (remoteDigest) {
      log('Firmware update available: cached=%s, remote=%s', cachedDigest, remoteDigest);
      onProgress?.('New firmware version available, pulling...');
    } else {
      log('Could not check remote digest, using cache');
      const size = statSync(paths.firmwareTar).size;
      onProgress?.(`Using cached firmware (${Math.round(size / 1024 / 1024)}MB)`);
      return paths.firmwareTar;
    }
  }

  // Pull image for ARM64
  log('Pulling firmware image: %s', image);
  onProgress?.('Pulling firmware image (this may take a few minutes)...');
  try {
    execSync(`docker pull --platform linux/arm64 "${image}"`, {
      stdio: 'pipe',
      timeout: 10 * 60 * 1000,
    });
  } catch {
    throw new Error('Failed to pull firmware image. Check that the image exists and GHCR credentials are correct.');
  }

  // Get the digest of what we just pulled
  const digest = getRemoteDigest(image);

  // Save to tar
  log('Saving firmware image to tar...');
  onProgress?.('Saving firmware image to disk...');
  try {
    execSync(`docker save "${image}" -o "${paths.firmwareTar}"`, {
      stdio: 'pipe',
      timeout: 5 * 60 * 1000,
    });
  } catch {
    throw new Error('Failed to save firmware image to disk');
  }

  // Write version tag + digest
  writeFileSync(paths.firmwareVersion, image);
  if (digest) {
    writeFileSync(paths.firmwareDigest, digest);
  }

  const size = statSync(paths.firmwareTar).size;
  log('Firmware image saved: %dMB (digest: %s)', Math.round(size / 1024 / 1024), digest || 'unknown');
  onProgress?.(`Firmware image ready (${Math.round(size / 1024 / 1024)}MB)`);
  return paths.firmwareTar;
}

export function clearFirmwareCache(): void {
  const paths = getCachePaths();
  try {
    execSync(`rm -f "${paths.firmwareTar}" "${paths.firmwareVersion}" "${paths.firmwareDigest}"`);
    log('Firmware cache cleared');
  } catch {
    // ignore
  }
}

File: backend/app/services/offline_assets.py
import logging
import os
import shutil
import subprocess

logger = logging.getLogger(__name__)

DOCKER_STATIC_URL = "https://download.docker.com/linux/static/stable/aarch64/docker-27.5.1.tgz"
CACHE_DIR = os.path.join(os.path.expanduser("~"), ".t3shield-installer")


def ensure_cache_dir():
    os.makedirs(CACHE_DIR, exist_ok=True)


def get_cache_paths() -> dict:
    return {
        "cache_dir": CACHE_DIR,
        "docker_tgz": os.path.join(CACHE_DIR, "docker-static.tgz"),
        "docker_dir": os.path.join(CACHE_DIR, "docker-static"),
        "firmware_tar": os.path.join(CACHE_DIR, "firmware.tar"),
        "firmware_version": os.path.join(CACHE_DIR, "firmware-version.txt"),
        "firmware_digest": os.path.join(CACHE_DIR, "firmware-digest.txt"),
    }


def get_cache_status() -> dict:
    paths = get_cache_paths()
    docker_ready = os.path.exists(os.path.join(paths["docker_dir"], "docker", "dockerd"))
    firmware_ready = os.path.exists(paths["firmware_tar"])
    firmware_tag = None
    if os.path.exists(paths["firmware_version"]):
        firmware_tag = open(paths["firmware_version"]).read().strip()
    return {"docker_binaries": docker_ready, "firmware_image": firmware_ready, "firmware_tag": firmware_tag}


def prepare_docker_binaries(on_progress=None) -> str:
    ensure_cache_dir()
    paths = get_cache_paths()

    dockerd = os.path.join(paths["docker_dir"], "docker", "dockerd")
    if os.path.exists(dockerd):
        logger.info("Docker binaries already cached")
        if on_progress:
            on_progress("Docker binaries cached")
        return paths["docker_dir"]

    if not os.path.exists(paths["docker_tgz"]):
        logger.info("Downloading Docker static binaries")
        if on_progress:
            on_progress("Downloading Docker binaries (60MB)...")
        subprocess.run(
            ["curl", "-fSL", "-o", paths["docker_tgz"], DOCKER_STATIC_URL],
            check=True, capture_output=True, timeout=300,
        )
        size_mb = os.path.getsize(paths["docker_tgz"]) // (1024 * 1024)
        logger.info("Downloaded: %dMB", size_mb)

    logger.info("Extracting Docker binaries")
    if on_progress:
        on_progress("Extracting Docker binaries...")
    os.makedirs(paths["docker_dir"], exist_ok=True)
    subprocess.run(
        ["tar", "xzf", paths["docker_tgz"], "-C", paths["docker_dir"]],
        check=True, capture_output=True,
    )

    if not os.path.exists(dockerd):
        raise RuntimeError("Docker extraction failed — dockerd not found")

    if on_progress:
        on_progress("Docker binaries ready")
    return paths["docker_dir"]


def _get_remote_digest(image: str) -> str | None:
    try:
        result = subprocess.run(
            ["docker", "manifest", "inspect", image],
            capture_output=True, text=True, timeout=30,
        )
        if result.returncode == 0:
            import json
            manifest = json.loads(result.stdout)
            return manifest.get("digest", "")
    except Exception:
        pass
    return None


def prepare_firmware_image(image: str, ghcr_username: str, ghcr_token: str, on_progress=None) -> str:
    ensure_cache_dir()
    paths = get_cache_paths()

    # Check docker is available
    if shutil.which("docker") is None:
        raise RuntimeError("Docker is not installed on this machine.")

    # Login to GHCR
    logger.info("Logging in to ghcr.io")
    if on_progress:
        on_progress("Logging in to container registry...")
    try:
        subprocess.run(
            f'echo "{ghcr_token}" | docker login ghcr.io -u "{ghcr_username}" --password-stdin',
            shell=True, check=True, capture_output=True, timeout=30,
        )
    except subprocess.CalledProcessError:
        raise RuntimeError("GHCR login failed — check username and token in Settings")

    # Check cache by digest
    if os.path.exists(paths["firmware_tar"]) and os.path.exists(paths["firmware_digest"]):
        cached_digest = open(paths["firmware_digest"]).read().strip()
        if on_progress:
            on_progress("Checking for firmware updates...")
        remote_digest = _get_remote_digest(image)
        if remote_digest and cached_digest == remote_digest:
            size_mb = os.path.getsize(paths["firmware_tar"]) // (1024 * 1024)
            logger.info("Firmware cache up to date (%dMB)", size_mb)
            if on_progress:
                on_progress(f"Firmware image up to date ({size_mb}MB)")
            return paths["firmware_tar"]
        if remote_digest:
            logger.info("New firmware available: cached=%s remote=%s", cached_digest, remote_digest)
        elif not remote_digest:
            size_mb = os.path.getsize(paths["firmware_tar"]) // (1024 * 1024)
            if on_progress:
                on_progress(f"Using cached firmware ({size_mb}MB)")
            return paths["firmware_tar"]

    # Pull
    logger.info("Pulling firmware image: %s", image)
    if on_progress:
        on_progress("Pulling firmware image...")
    try:
        subprocess.run(
            ["docker", "pull", "--platform", "linux/arm64", image],
            check=True, capture_output=True, timeout=600,
        )
    except subprocess.CalledProcessError:
        raise RuntimeError("Failed to pull firmware image — check credentials and image name")

    digest = _get_remote_digest(image)

    # Save
    logger.info("Saving firmware to tar")
    if on_progress:
        on_progress("Saving firmware image to disk...")
    try:
        subprocess.run(
            ["docker", "save", image, "-o", paths["firmware_tar"]],
            check=True, capture_output=True, timeout=300,
        )
    except subprocess.CalledProcessError:
        raise RuntimeError("Failed to save firmware image to disk")

    with open(paths["firmware_version"], "w") as f:
        f.write(image)
    if digest:
        with open(paths["firmware_digest"], "w") as f:
            f.write(digest)

    size_mb = os.path.getsize(paths["firmware_tar"]) // (1024 * 1024)
    logger.info("Firmware saved: %dMB", size_mb)
    if on_progress:
        on_progress(f"Firmware image ready ({size_mb}MB)")
    return paths["firmware_tar"]


def clear_firmware_cache():
    paths = get_cache_paths()
    for key in ("firmware_tar", "firmware_version", "firmware_digest"):
        if os.path.exists(paths[key]):
            os.remove(paths[key])
    logger.info("Firmware cache cleared")
